lib/active_merchant/billing/gateways/square.rb

```diff
@@ -0,0 +1,283 @@
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    class SquareGateway < Gateway
+      self.test_url = 'https://connect.squareupsandbox.com/v2'
+      self.live_url = 'https://connect.squareup.com/v2'
+
+      self.supported_countries = ['US', 'CA', 'GB', 'AU', 'JP']
+      self.default_currency = 'USD'
+      self.supported_cardtypes = [:visa, :master, :american_express, :discover, :jcb, :union_pay]
+      self.money_format = :cents
+
+      self.homepage_url = 'https://squareup.com/'
+      self.display_name = 'Square Payments Gateway'
+
+      DEFAULT_API_VERSION = '2019-10-23'
+
+      STANDARD_ERROR_CODE_MAPPING = {
+        'BAD_EXPIRATION' => STANDARD_ERROR_CODE[:invalid_expiry_date],
+        'INVALID_ACCOUNT' => STANDARD_ERROR_CODE[:config_error],
+        'CARDHOLDER_INSUFFICIENT_PERMISSIONS' => STANDARD_ERROR_CODE[:card_declined],
+        'INSUFFICIENT_PERMISSIONS' => STANDARD_ERROR_CODE[:config_error],
+        'INSUFFICIENT_FUNDS' => STANDARD_ERROR_CODE[:card_declined],
+        'INVALID_LOCATION' => STANDARD_ERROR_CODE[:processing_error],
+        'TRANSACTION_LIMIT' => STANDARD_ERROR_CODE[:card_declined],
+        'CARD_EXPIRED' => STANDARD_ERROR_CODE[:invalid_expiry_date],
+        'CVV_FAILURE' => STANDARD_ERROR_CODE[:card_declined],
+        'ADDRESS_VERIFICATION_FAILURE' => STANDARD_ERROR_CODE[:processing_error],
+        'VOICE_FAILURE' => STANDARD_ERROR_CODE[:card_declined],
+        'PAN_FAILURE' => STANDARD_ERROR_CODE[:incorrect_number],
+        'EXPIRATION_FAILURE' => STANDARD_ERROR_CODE[:invalid_expiry_date],
+        'CARD_NOT_SUPPORTED' => STANDARD_ERROR_CODE[:processing_error],
+        'INVALID_PIN' => STANDARD_ERROR_CODE[:incorrect_pin],
+        'INVALID_POSTAL_CODE' => STANDARD_ERROR_CODE[:incorrect_zip],
+        'CHIP_INSERTION_REQUIRED' => STANDARD_ERROR_CODE[:processing_error],
+        'ALLOWABLE_PIN_TRIES_EXCEEDED' => STANDARD_ERROR_CODE[:card_declined],
+        'MANUALLY_ENTERED_PAYMENT_NOT_SUPPORTED' => STANDARD_ERROR_CODE[:unsupported_feature],
+        'PAYMENT_LIMIT_EXCEEDED' => STANDARD_ERROR_CODE[:processing_error],
+        'GENERIC_DECLINE' => STANDARD_ERROR_CODE[:card_declined],
+        'INVALID_FEES' => STANDARD_ERROR_CODE[:config_error],
+        'GIFT_CARD_AVAILABLE_AMOUNT' => STANDARD_ERROR_CODE[:card_declined],
+        'BAD_REQUEST' => STANDARD_ERROR_CODE[:processing_error]
+      }
+
+      def initialize(options={})
+        requires!(options, :access_token)
+        @access_token = options[:access_token]
+        @fee_currency = options[:fee_currency] || default_currency
+        super
+      end
+
+      def authorize(money, payment, options={})
+        post = create_post_for_auth_or_purchase(money, payment, options)
+        post[:autocomplete] = false
+
+        commit(:post, 'payments', post, options)
+      end
+
+      def purchase(money, payment, options={})
+        post = create_post_for_auth_or_purchase(money, payment, options)
```

Comment:
It might be nice to add `requires!(options, :idempotency_key)` for all these instead of sending them to the gateway.